scipy/stats/_censored_data.py

```diff
@@ -0,0 +1,192 @@
+        raise ValueError('x must not contain nan or inf.')
+    if censored.size != x.size:
+        raise ValueError('x and censored must have the same length.')
+    return x, censored.astype(bool)
+
+
+class CensoredData:
+    """
+    Instances of this class represent censored data.
+
+    Instances may be passed to the ``fit`` method of continuous
+    univariate SciPy distributions for MLE fitting.
+
+    Left-, right-, and interval-censored data can be represented.
+
+    For interval-censored data, pass the lower and upper limits of
+    the intervals as one-dimensional arrays to the ``CensoredData``
+    constuctor.  By using the values `np.inf` and `-np.inf`, left-
+    and right-censored data may also be created with the constructor.
+
+    For convenience, the class methods ``left_censored`` and
+    ``right_censored`` are provided to create a ``CensoredData``
+    instance from a single one-dimensional array of observations
+    and a corresponding boolean array to indicate which observations
+    are censored.
+
+    Examples
+    --------
+    In the most general case, a censored data set may contain values that
+    are left-censored, right-censored, interval-censored and not censored.
+    For example, here we create a data set with length five:
+
+    >>> from scipy.stats import CensoredData
+    >>> data = CensoredData([-np.inf, 1, 1.5, 2, 10],
+    ...                     [0,       1, 1.5, 3, np.inf])
+    >>> print(data)
+    CensoredData(5 values: 2 not censored, 1 left-censored,
+    1 right-censored, 1 interval-censored)
+
+    The first value is left-censored; the true (but unknown) value is
+    less than or equal to 0.  The second and third values, 1 and 1.5,
+    are not censored.  The fourth value is interval-censored; the true
+    value is in the interval [2, 3].  The last value is right-censored;
+    the true value is greater than or equal to 10.
+
+    A common case is to have data that is a mix of uncensored values and
+    censored values that are all right-censored (or all left-censored)
+    with the same bound on the censored values.  For example, a measuring
+    device might have an upper limit of 100, so a reading of 100 means
+    the true value is unknown, but the value is not less than 100.
+    Suppose we have nine readings from such a device, with six being
+    [65, 51, 88, 93, 96, 89] and three being 100.  To create an instance
+    of `CensoredData` to represent this data, we can use the class method
+    `CensoredData.right_censored` as follows:
+
+    >>> values = [65, 51, 88, 93, 96, 89, 100, 100, 100]
+    >>> censored = [0, 0, 0, 0, 0, 0, 1, 1, 1]
+
+    A 1 (or any True value) in the sequence ``censored`` indicates
+    that the corresponding value in ``values`` is right-censored.
+
+    >>> data = CensoredData.right_censored(values, censored)
+    >>> print(data)
+    CensoredData(9 values: 6 not censored, 3 right-censored)
+    """
+
+    def __init__(self, lower, upper):
```

Comment:
Why is there no `data` input here? How would you go from using the constructor to adding data later to get the equivalent of the `right_censored` classmethod?

Reply:
All the data is given in `lower` and `upper`.  I hope the explanation in the new `Parameters` section clarifies the meaning of these arrays.

To create right-censored data in the constructor, put `inf` in `upper` to indicate that the upper end of the interval is infinite.  E.g. `lower=[12, 25, 100], upper=[12, 25, np.inf]` represents three measurements.  The first two are not censored.  The third is right-censored (the true but unknown value is greater than 100).

Reply:
This example won't work, which demonstrates a problem with the usage of `inf`. `lower=[12, 25, 100]` will be turned into an integer array, `upper=[12, 25, np.inf]` into a floating point array. The `==` for floating point numbers is something we usually recommend against, and here could result in corner cases where the comparison isn't always true when putting the same value through different casting paths.

Reply:
Why not `def __init__(self, x, lower, upper):`? Seems easier to understand, and works for all dtypes.

Reply:
The dtype can be dealt with.  There are many other possible APIs, though, so I'm happy to consider alternatives.  What if we took your suggestion and expanded it to include interval-censored values as an additional argument?  E.g. `def __init__(self, x=None, left=None, right=None, intervals=None)`, where `x` is a 1D array of uncensored values, `left` is a 1D array of bounds for left-censored data, `right` is a 1D array of bounds for right-censored data, and `intervals` is a 2D array with shape (m, 2) containing the intervals for `m` interval-censored values?  (It might be tough to later add the `count` parameter in a nice way.)

Should we move this API design discussion to the mailing list for greater visibility?

Reply:
> Should we move this API design discussion to the mailing list for greater visibility?

I think it'd be good to outline a number of options (maybe in comments to the PR rather than in the diff, or even better in a separate issue specifically for the API of `CensoredData`), and point that out on the mailing list. More discoverable in the future, and easier to follow with markdown formatting.

Also would help to point at how R does this.

>  E.g. `def __init__(self, x=None, left=None, right=None, intervals=None), `

`x` should be required I think? To me one obvious issue that needs fixing is that the classmethods are not a subset of the main constructor. `x` should have the same meaning in all of those. The classmethods get it right imho, a simple parameter `x` that is the data.

Reply:
> x should be required I think?

In the API with `x`, `left`, `right` and `intervals`, `x` holds only the values that are not censored.  Since it is possible to have a data set in which all the values are censored, `x` is optional.

For clarity, here's how the two examples in the CensoredData docstring would look when the constructor API is `def __init__(self, x=None, left=None, right=None, intervals=None)`.  (Note that with this API, the class methods `left_censored` and `right_censored`--while still potentially convenient for the user--aren't as big a benefit as they were with the old `(lower, upper)` API.)

    >>> data = CensoredData(x=[1, 1.5], left=[0], right=[10], intervals=[[2, 3]])

    >>> data = CensoredData(x=[13, 22, 17, 15], right=[20, 18])

What I like about this version of the API is that the specification of each censoring type is explicit. The user never has to fill in ±inf anywhere, and instead just says, in effect, "these are the regular data points, these are the right-censored data points," etc.

>  ... even better in a separate issue specifically for the API of CensoredData

👍

Reply:
This is a useful conversation, so leaving it unresolved. The API is now much clearer, I'm happy with this.